### PR TITLE
WIP: MEEG data

### LINKED ISSUE #4: ENH: Add support for (M/EEG) FIFF file format?
(Copied this issue from the old repo, where I errantly posted it.)

I haven't used NIMS yet, but my group is considering adopting NIMS for (at least) MRI data organization. In the coming months we plan on trying to set up NIMS and see if it suits our data management needs.

In our group we also do M/EEG studies. It seems like being able to use NIMS as a front-end for MEG data organization would be really cool. I talked with @jyeatman about the possibility today, and he said that you folks might be receptive to this idea. He told me I'd need a way to scrape the header information from MEG files. Assuming [mne-python](https://github.com/mne-tools/mne-python) (I am a dev of the package) is installed this is a two-liner, so I don't think it would be too difficult:

```
import mne
info_dict = mne.io.read_info(fname)
```

`info_dict` then has necessary information to identify the scan, including the subject name, scan date, etc. Assuming this worked, it would also make a tight integration of mne-python pipelines with NIMS possible (cc @agramfort).

Let me know if this is something you would potentially be interested in having as part of NIMS!

### PR DESCRIPTION
**Pull request by Eric89GXL**
Currently in the state that it will store the data in `persistent/data/` when used with https://github.com/scitran/reaper/pull/12 and https://github.com/scitran/scitran/pull/57.

Closes #4.

### CODE REVIEW

**Comment by Eric89GXL** [comment]
How do I see the data that I've added? I click the "login" button, but a window pops up and immediately disappears. I assume I need to actually add a user to the system somewhere...?

Also, in the code I have a lot of `XXX`s, which indicate that I don't know how to set those variables appropriately. Are the meanings of the `nims_` parameters documented somewhere? I was hoping they'd be documented under `class Reader` but there isn't anything there yet. I'm happy to add some docstrings to all those properties if someone can enumerate them, or we can video-chat about them. I am curious what I should do when a property isn't defined, though. I am also not sure how to populate the entries for the `session_properties` that are created.

**Comment by gsfr** [comment]
Since Kevin has left our team, this repo has been largely unmaintained. I will add line comments to your code regarding the ```nims_``` properties.
When you first run ```live.sh configure```, it should ask you about bootstrapping a user. You need to provide a Google OAuth capable email address. With that, you will be able to log in.
Also, please fix the CI errors.

File: scitran/data/meeg/meeg.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+            raw.preload_data()
+
+    @property
+    def nims_project(self):
+        return self.project_name
+
+    @property
+    def nims_group_id(self):
+        # XXX
+        return 'group_id'
```
Comment (gsfr) [inline scitran/data/meeg/meeg.py]: SciTran data are organized in a group, project, session, acquisition hierarchy. A group is approximately equivalent to a university research lab. The group "owns" the project. If the provided group does not exist, data will end up in the "unknown" group. By default, ```live.sh``` provisions a "scitran" group.

**Comment by Eric89GXL** [comment]
I provided my gmail address during the configure step, but when I click "login" a window quickly appears and disappears. Not sure what the deal is with that.

I assumed the CI errors were due to `mne` not being installed in `scitran` (https://github.com/scitran/scitran/pull/57), but I suppose I also need to add an installer to the Travis script here, as well. I'll do that now.

File: scitran/data/meeg/meeg.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+    @property
+    def nims_session_id(self):
+        # XXX
+        return 'session_id'
```
Comment (gsfr) [inline scitran/data/meeg/meeg.py]: This is a unique ID for the session that contains this acquisition. A session is roughly equivalent to a DICOM study. For DICOM we use the study instance UID.

File: scitran/data/meeg/meeg.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+    @property
+    def nims_session_label(self):
+        # XXX
+        return 'session_label'
```
Comment (gsfr) [inline scitran/data/meeg/meeg.py]: A human-readable label to identify the session.

**Comment by Eric89GXL** [comment]
Actually, we're going to run into the same problem here as we did in https://github.com/scitran/scitran/pull/57, namely that I need to install `scipy` and it's going to be too slow here. I can change Travis to use Anaconda, or let it use the system packages and install `numpy` and `scipy` from there. Will either of those options work here? In theory using the system packages should make the build quite a bit faster.

File: scitran/data/meeg/meeg.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+    @property
+    def nims_session_subject(self):
+        # XXX
+        return 'subj_code'
```
Comment (gsfr) [inline scitran/data/meeg/meeg.py]: An identifier for the subject. For DICOM, we get this from the patient ID.

File: scitran/data/meeg/meeg.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+    @property
+    def nims_acquisition_label(self):
+        # XXX
+        return 'series_no'
```
Comment (gsfr) [inline scitran/data/meeg/meeg.py]: A human-readable label to identify the acquisition. A timestamp works, if nothing else is available.

File: scitran/data/meeg/meeg.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+
+    @property
+    def nims_acquisition_description(self):
+        # XXX get from info
+        return 'foobar'
```
Comment (gsfr) [inline scitran/data/meeg/meeg.py]: This should come from your data. Maybe just "M/EEG" for now.

File: scitran/data/meeg/meeg.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+
+    @property
+    def nims_file_name(self):
+        # XXX what are the constraints on this?
+        return 'foo'
```
Comment (gsfr) [inline scitran/data/meeg/meeg.py]: What should be call this file on disk? Globally unique names are nice, but not required. We store all files for an acquisition in the same directory, so some uniqueness is needed.

File: scitran/data/meeg/meeg.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+
+    @property
+    def nims_file_kinds(self):
+        # XXX
+        return ['FIF']
```
Comment (gsfr) [inline scitran/data/meeg/meeg.py]: This is a weird one. We are working on a simplified [data model](https://github.com/scitran/api/wiki/Data-Model).

File: scitran/data/meeg/meeg.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+    @property
+    def nims_acquisition_id(self):
+        # XXX
+        return 'nims_acquisition_id'
```
Comment (gsfr) [inline scitran/data/meeg/meeg.py]: This is a unique ID for this acquisition. An acquisition is roughly equivalent to a DICOM series. For DICOM we use the series instance UID.

File: scitran/data/meeg/meeg.py [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+
+    # XXX why are these not implemented at the data.Reader level?
```
Comment (Eric89GXL) [inline scitran/data/meeg/meeg.py]: @gsfr any reason not to have the following as part of the super `reader.Reader` class? It would make the code more DRY to have these as defaults in that class. But if you want `Reader` to be an ABC I guess it makes sense.

**Comment by gsfr** [comment]
That is probably the Google login window that comes and goes. Please try using Chrome with the default popup blocking settings.

File: scitran/data/meeg/meeg.py [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+            'lastname': {
+                'field': 'subj_lastname',
+                'title': 'Last Name',
+                'type': 'string',
+            },
+            'dob': {
+                'field': 'subj_dob',
+                'title': 'Date of Birth',
+                'type': 'string',
+                'format': 'date',
```
Comment (Eric89GXL) [inline scitran/data/meeg/meeg.py]: @gsfr how / where do I populate these fields? I looked at DICOM, and at least to start, I know I can fill these fields based on what's in the MEG files.

**Comment by Eric89GXL** [comment]
That's what I'm on (Chrome w/defaults) I think. I actually see the window open up and close, so it's not getting blocked. And If I have the webpage open, and kill the server, then the window opens with a URL and a not-found error as you'd expect -- but it at least indicates that, when the server is running, it is visiting the appropriate URL and auto-closes really quickly. Is that normal behavior? Even after that there is no indication that I am logged in, the "Login" link is still there, and clicking it still makes the window pop up and disappear.

**Comment by Eric89GXL** [comment]
FYI Travis build time dropped from 4 minutes to 1:42 with the use of the `scikit-image` wheels to install dependencies, despite the fact that I added two (`scipy` and `mne`)

File: scitran/data/meeg/meeg.py (hunk 10 above)
Comment (Eric89GXL) [inline scitran/data/meeg/meeg.py]: Although already `Reader` isn't purely ABC because it has e.g. the `__str__` method, so it might be worth not setting the below values to be abstract. WDYT?

File: scitran/data/meeg/meeg.py (hunk 11 above)
Comment (Eric89GXL) [inline scitran/data/meeg/meeg.py]: ahhh, based on looking at other modules, I think I need to set the e.g. `.subj_firstname` attribute of the `Reader` class.

File: scitran/data/meeg/meeg.py (hunk 10 above)
Comment (gsfr) [inline scitran/data/meeg/meeg.py]: I think that would be fine, but I'm reluctant to make unnecessary changes while this repo does not have a maintainer.

**Comment by gsfr** [comment]
Sounds like you have CI figured out. I think the rules are pretty loose in this repo, given the lack of a maintainer.

I really don't know what's going on with the login window. Have you looked at Chrome's console?
Come to think of it, you might have to edit the ```BASE_URL``` in ```sdm/app/custom/config.js```. It is still hardcoded and does not use the newish scitran config system.

@rentzso Could you please upgrade SDM so that a fresh install via ```live.sh``` works out of the box?

**Comment by rentzso** [comment]
@gsfr done with this commit: https://github.com/scitran/sdm/commit/fdbaba40bc8b95f4b38ba3fb844c989c30a13eeb

**Comment by Eric89GXL** [comment]
Closing for PR into `master` instead of `zip_file`
